Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift

```diff
@@ -119,25 +131,322 @@ public class AesKw {
 
 }  // AesKw
 
+public enum HashFunction {
+    case sha256
+    case sha384
+    case sha512
+}
+
 public class Digest {
     public static func sha1(_ data: SpanConstUInt8) -> VectorUInt8 {
-        return digest(data, Insecure.SHA1.self)
+        return digest(data, t: Insecure.SHA1.self)
     }
     public static func sha256(_ data: SpanConstUInt8) -> VectorUInt8 {
-        return digest(data, SHA256.self)
+        return digest(data, t: SHA256.self)
     }
     public static func sha384(_ data: SpanConstUInt8) -> VectorUInt8 {
-        return digest(data, SHA384.self)
+        return digest(data, t: SHA384.self)
     }
     public static func sha512(_ data: SpanConstUInt8) -> VectorUInt8 {
-        return digest(data, SHA512.self)
+        return digest(data, t: SHA512.self)
     }
-    private static func digest<T: HashFunction>(_ data: SpanConstUInt8, _: T.Type)
-        -> VectorUInt8
+    fileprivate static func digest<T: CryptoKit.HashFunction>(_ data: SpanConstUInt8, _: T.Type)
+        -> T.Digest
     {
         var hasher = T()
         hasher.update(data: data)
-        return hasher.finalize().copyToVectorUInt8()
+        return hasher.finalize()
+    }
+
+    fileprivate static func digest<T: CryptoKit.HashFunction>(_ data: SpanConstUInt8, t: T.Type)
+        -> VectorUInt8
+    {
+        return Self.digest(data, t).copyToVectorUInt8()
+    }
+
+    fileprivate static func digest(_ data: SpanConstUInt8, hashFunction: HashFunction)
+        -> any CryptoKit.Digest
+    {
+        switch hashFunction {
+        case .sha256:
+            return digest(data, SHA256.self)
+        case .sha384:
+            return digest(data, SHA384.self)
+        case .sha512:
+            return digest(data, SHA512.self)
+        }
+    }
```

Comment:
I think that `any CryptoKit.Digest` will incur a performance penalty, because now you have to box the digest type in an existential container, which may move it to the heap. Perhaps you can delete this function and move the switch statement to call sites?

Ideally, we could remove the `HashFunction` type altogether, since using it instead of a generic parameter loses type information. But I'm not sure whether it's possible to, say, replace `sign(message:,hashFunction:)` with a generic version.

Reply:
Without actually measuring the impact, there is a performance penalty for boxing and table lookup, which is probably a minion as compared to the time spent doing the actual crypto.

The sign method as designed is dependent the hash parameter specified at the JS layer which gets passed down. The HashFunction Type here is used to bridge that information down so we need it at the interface layer.
Calling a generic `sign(message:hashFunction)` from C++ is not really possible today I think so we have to take that parameter as input.
But I see your point. I will try to find another solution if possible which is not duplicating the switch case everywhere.. cause that has the risk of having higher maintenance cost and more lines of code.